types.nix

```
@@ -0,0 +1,278 @@
+{ lib ? import <nixpkgs/lib> }:
+with lib;
+with builtins;
+
+rec {
+
+  diskoLib = {
+    subType = typeAttr: mkOptionType rec {
+      name = "subType";
+      description = "one of ${attrNames typeAttr}";
+      check = x: typeAttr.${x.type}.check x;
+      merge = loc: defs:
+        map (def: typeAttr.${def.value.type}.merge loc [def]) defs;
+      nestedTypes = typeAttr;
+    };
+    partitionType = mkOption {
+      # not sure if nested tables are a thing? maybe we need an extra type for table content
+      type = types.nullOr (diskoLib.subType { inherit table btrfs filesystem zfs mdraid luks lvm_pv; });
+      default = null;
+    };
+  };
+
+  btrfs = types.submodule {
+    options = {
+      type = mkOption {
+        type = types.enum [ "btrfs" ];
+        internal = true;
+      };
+      subvolumes = mkOption {
+        type = types.listOf types.str; # TODO path type
+        default = [];
+      };
+      mountpoint = mkOption {
+        type = types.str; # TODO absolute-path type;
+      };
+    };
+  };
+
+  filesystem = types.submodule {
+    options = {
+      type = mkOption {
+        type = types.enum [ "filesystem" ];
+        internal = true;
+      };
+      extraArgs = mkOption {
+        type = types.str;
+        default = "";
+      };
+      options = mkOption {
+        type = types.listOf types.str;
+        default = [];
+      };
+      mountpoint = mkOption {
+        type = types.str; # TODO absolute-path
+      };
+      format = mkOption {
+        type = types.enum [ "ext1" "ext2" "ext3" "ext4" "btrfs" "xfs" ]; # TODO add many more
```

Not sure if this should be type checked at all. It's kind of open-ended and potentially user extensible:

```suggestion
        type = types.enum [ "ext1" "ext2" "ext3" "ext4" "btrfs" "xfs" "vfat" "f2fs" ]; # TODO add many more
```

And this is still missing bcachefs, jfs and others...

---

hmm yeah we can leave that probably open and just let the script fail if the binary is missing